stan/math/prim/scal/prob/std_normal_lpdf.hpp

```diff
@@ -0,0 +1,75 @@
+#ifndef STAN_MATH_PRIM_SCAL_PROB_STD_NORMAL_LPDF_HPP
+#define STAN_MATH_PRIM_SCAL_PROB_STD_NORMAL_LPDF_HPP
+
+#include <stan/math/prim/scal/meta/is_constant_struct.hpp>
+#include <stan/math/prim/scal/meta/partials_return_type.hpp>
+#include <stan/math/prim/scal/meta/operands_and_partials.hpp>
+#include <stan/math/prim/scal/meta/scalar_seq_view.hpp>
+#include <stan/math/prim/scal/err/check_not_nan.hpp>
+#include <stan/math/prim/scal/fun/constants.hpp>
+#include <stan/math/prim/scal/fun/value_of.hpp>
+#include <stan/math/prim/scal/meta/include_summand.hpp>
+#include <boost/random/normal_distribution.hpp>
+#include <boost/random/variate_generator.hpp>
+#include <string>
+
+namespace stan {
+  namespace math {
+
+    /**
+     * The log of the normal density for the specified scalar(s) given
+     * a mean of 0 and a standard deviation of 1. y can be either
+     * a scalar or a vector.
+     *
+     * <p>The result log probability is defined to be the sum of the
+     * log probabilities for each observation.
+     * @tparam T_y Underlying type of scalar in sequence.
+     * @param y (Sequence of) scalar(s).
+     * @return The log of the product of the densities.
+     * @throw std::domain_error if any scalar is nan.
+     */
+    template <bool propto, typename T_y>
+    typename return_type<T_y>::type std_normal_lpdf(const T_y& y) {
+      static const std::string function = "std_normal_lpdf";
+      typedef typename stan::partials_return_type<T_y>::type
+        T_partials_return;
+
+      using stan::is_constant_struct;
+
+      if (!(stan::length(y)))
+        return 0.0;
+
+      T_partials_return logp(0.0);
+
+      check_not_nan(function, "Random variable", y);
+      if (!include_summand<propto, T_y>::value)
+        return 0.0;
+
+      operands_and_partials<T_y> ops_partials(y);
+
+      scalar_seq_view<T_y> y_vec(y);
+
+      for (size_t n = 0; n < length(y); n++) {
+        const T_partials_return y_dbl = value_of(y_vec[n]);
+
+        static double NEGATIVE_HALF = -0.5;
+
+        if (include_summand<propto>::value)
+          logp += NEG_LOG_SQRT_TWO_PI;
+        if (include_summand<propto, T_y>::value)
```

Comment:
It's worth doing this case (and fixing the regular normal) to do this loop more efficiently.  It will be more efficient if written as (I also brought down the declaration of `logp` to just before it's used).

```
T_partials_return logp(0.0);
if (include_summand<propto, T_y>::value) {  // same condition as !is_constant_struct<T_y>::value
  for (size_t n = 0; n < length(y); n++) {
    T_partials_return y_val = value_of(y_vec[n]);
    logp += y_val * y_val;
    ops_partials.edge1_.partials_[n] -= y_val; 
  }
  logp *= -0.5;  // only multiply once
}
if (include_summand<propto>::value)
  logp += length(y) * NEG_LOG_SQRT_TWO_PI;  // replace additions with one multiplication
```

If `length(y) == 1`, the second step is a little less efficient than just adding the single value.

It might even be possible to do this all with vector operations if we can map y_vec.

It would probably be even faster just to split this on all the possible inputs, as it's just `y`.  That is, have an implementation where `y` is scalar, another where `y` is a vector or row vector (dot product for log density and vector ops for derivatives), and yet another for where it's an array that maps it using `Eigen::Map`.

Anyway, this isn't all necessary here, but I'd very much like to start making these low-level implementations that get used all over the place as efficient as possible.  And this one may wind up being our most highly used one, so if you don't want to do this efficiency change, could you please take this comment and make an issue out of it and I can do it, or someone else can in another pass.

Reply:
I can do this for `std_normal_lpdf`. If we split it on inputs, what if someone tries to pass in a matrix for `y`?